components/webgpu/lib.rs

```diff
@@ -1146,50 +1111,55 @@ impl<'a> WGPU<'a> {
                             &[encoder_id.transmute()]
                         ));
 
-                        let map_info = BufferMapInfo {
-                            buffer_id,
-                            sender: self.sender.clone(),
-                            global: &self.global,
-                            size: buffer_size as usize,
-                            external_id: Some(external_id),
-                        };
-                        self.present_buffer_maps
-                            .insert(buffer_id, Rc::new(map_info));
-                        // TODO(sagudev): replace with safe callback
-                        unsafe extern "C" fn callback(
-                            status: BufferMapAsyncStatus,
-                            userdata: *mut u8,
-                        ) {
-                            let info = Rc::from_raw(
-                                userdata
-                                    as *const BufferMapInfo<(Option<ErrorScopeId>, WebGPURequest)>,
-                            );
-                            match status {
-                                BufferMapAsyncStatus::Success => {
-                                    if let Err(e) = info.sender.send((
-                                        None,
-                                        WebGPURequest::UpdateWebRenderData {
-                                            buffer_id: info.buffer_id,
-                                            buffer_size: info.size,
-                                            external_id: info.external_id.unwrap(),
-                                        },
-                                    )) {
-                                        warn!("Could not send UpdateWebRenderData ({})", e);
+                        let glob = Arc::clone(&self.global);
+                        let wgpu_image_map = Arc::clone(&self.wgpu_image_map);
+                        let webrender_api = Arc::clone(&self.webrender_api);
+                        let webrender_document = self.webrender_document;
+                        let callback = BufferMapCallback::from_rust(Box::from(move |result| {
```

Comment:
Can we not just call `global.buffer_map_async` synchronously, and then handle the result and do the webrender stuff?